File: solutions/133/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Sat Dec 18 15:58:20 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <iostream> 
#include <memory> 
#include <unordered_map> 
#include <gtest/gtest.h> 

using namespace std; 

template <typename T> 
class Node{
public:
    T val; 
    vector<shared_ptr<Node<T>>> neighbors; 
    Node() : val(0) {

    }
    Node(T _val) : val(_val) {

    }
    Node(T _val, vector<shared_ptr<Node<T>>> _neighbors) : val(_val), neighbors(_neighbors) {

    }
};

template <typename T> 
class Solution {
private:
    // old node, cloned node
    unordered_map<shared_ptr<Node<T>>, shared_ptr<Node<T>>> hash;
public:
    shared_ptr<Node<T>> dfs(shared_ptr<Node<T>> node) {
        // base case, return the node if already cloned
        if (hash.find(node) != hash.end())
            return hash[node];
        // else, create the clone and clone edges
        auto nodeClone = make_shared<Node<T>>(node->val);
        hash[node] = nodeClone;
        for (auto& n : node->neighbors)
            nodeClone->neighbors.push_back(dfs(n));
        return hash[node];
    }
    shared_ptr<Node<T>> cloneGraph(shared_ptr<Node<T>> node) {
        if (node == nullptr) return nullptr;
        return dfs(node);
    }
};

TEST(Test133, SimpleTest) {
    auto node1 = make_shared<Node<int>>(1); 
    auto node2 = make_shared<Node<int>>(2); 
    auto node3 = make_shared<Node<int>>(3); 
    auto node4 = make_shared<Node<int>>(4); 
    node1->neighbors.push_back(node2);
    node1->neighbors.push_back(node4);
    node2->neighbors.push_back(node1);
    node2->neighbors.push_back(node3);
    node3->neighbors.push_back(node2);
    node3->neighbors.push_back(node4);
    node4->neighbors.push_back(node1);
    node4->neighbors.push_back(node3);

    Solution<int> s; 
    auto nodeClone = s.cloneGraph(node1); 
    for (auto& n : nodeClone->neighbors) 
        cout << n->val << " ";
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: solutions/133/main.py
from typing import List 

# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if node is None:
            return None

        gQue = []
        gQue.append(node)
        visited = []
        visited.append(node)

        nodes_clone = {}

        while len(gQue) > 0:
            node_old = gQue[0]
            gQue.pop(0)
            if node_old.val not in nodes_clone:
                nodes_clone[node_old.val] = Node(node_old.val)
            node_new = nodes_clone[node_old.val]
            for n in node_old.neighbors:
                if n.val not in nodes_clone:
                    nodes_clone[n.val] = Node(n.val)
                n_clone = nodes_clone[n.val]
                node_new.neighbors.append(n_clone)
                if n not in visited:
                    gQue.append(n)
                    visited.append(n)

        return nodes_clone[node.val]

if __name__ == "__main__": 
    node1 = Node(1)
    node2 = Node(2)
    node3 = Node(3)
    node4 = Node(4)
    node1.neighbors.append(node2)
    node1.neighbors.append(node4)
    node2.neighbors.append(node1)
    node2.neighbors.append(node3)
    node3.neighbors.append(node2)
    node3.neighbors.append(node4)
    node4.neighbors.append(node1)
    node4.neighbors.append(node3)

    s = Solution()
    node1_clone = s.cloneGraph(node1)
    for n in node1_clone.neighbors: 
        print(n.val)

